New Feature: Food Search 2.0 for Loop!

taylorpatterson-T1D opened the pull request:
This PR has been deprecated and replaced with PR #2404 

Introducing a new AI-enabled food search capability for Loop. Loop has helped me become a better diabetic and I would love to give back to the community by contributing this new feature for Loop, which I believe will benefit all Loop users.

**PROBLEM WE'RE SOLVING:** Diabetics often lack nutrition knowledge when estimating carb intake. And we need that knowledge when we’re making dosing decisions, especially when eating out at restaurants - this knowledge is the single most important tool we have to improve our Time in Range and A1C. Loop's Food Search feature uses barcodes and AI analysis to provide accurate nutrition information and advanced diabetes management recommendations.

**NEW FEATURE'S IMPACT:** The Food Search system is a comprehensive food analysis and nutrition tracing solution integrated into Loop for improved diabetes management. It provides multiple search options including text, barcode scanning, and AI-powered image analysis that produces comprehensive nutrition data and dosing suggestions via diabetic notes. With Food Search, we can enter a meal bolus with **MUCH greater accuracy** which should improve Time in Range and A1C. The AI-analysis engine can even read menus (in multiple languages) and provide guidance for dosing.

Short video demo: https://youtu.be/t7JxvkGcMfU

Review the docs in “/Loop/Documentation/FoodSearch 2.0 Docs/” for comprehensive documentation for Loop's Food Search functionality, including AI-powered nutrition analysis and advanced diabetes management recommendations. API setup guides are provided so the user can enter their personal AI API keys to enable AI food analysis. 

This new PR (Food-Search 2.0) fixes compatibility bugs from the initial commit and adds Advanced AI Analysis including FPU analysis, and Dynamic Dosing recommendations.

**Requesting a review from the maintainers? @ps2, @marionbarker, @loudnate**


Feature is hidden behind a Feature Flag in **Loop > Settings > Enable Food Search**:
<img width="248" height="547" alt="Screenshot 2025-07-20 at 4 02 05 PM" src="https://github.com/user-attachments/assets/dd329adf-db3b-4d01-be87-528d9f48ff9e" />

New **Search for Food** section on the Add Carb Entry view:
<img width="249" height="547" alt="Screenshot 2025-07-20 at 4 02 18 PM" src="https://github.com/user-attachments/assets/0d8b412a-0fd8-4792-b769-006b8a4fb0b0" />

Simple text search:
<img width="250" height="547" alt="Screenshot 2025-07-20 at 4 02 30 PM" src="https://github.com/user-attachments/assets/753ec4d7-0afa-4483-914b-e28f122d702c" />

Barcode scan search for high accuracy of packaged foods:
<img width="252" height="549" alt="Screenshot 2025-07-20 at 4 02 40 PM" src="https://github.com/user-attachments/assets/ed43a576-5db9-4007-8d6e-61beb871500f" />

AI-enabled food image analysis engine in action:
<img width="251" height="547" alt="Screenshot 2025-07-20 at 4 02 52 PM" src="https://github.com/user-attachments/assets/1db84f07-8f27-46d5-81a1-ae1b7e49ced7" />

**AI-Analysis** results with comprehensive analysis details including an item by item breakdown of the meal:
<img width="253" height="546" alt="Screenshot 2025-07-20 at 4 03 05 PM" src="https://github.com/user-attachments/assets/7069aba9-4d61-4075-99ce-3d8bbe9abc30" />

Even more detail about how **Portions & Servings** were calculated and **Diabetic Notes** to advise during dosing:
<img width="252" height="546" alt="Screenshot 2025-07-20 at 4 03 22 PM" src="https://github.com/user-attachments/assets/f0f3298b-0618-4982-82d6-4fe2ce770cca" />

Advanced AI Analysis Mode for those who want to geek out on **FPU** and **Advanced Dosing Recommendations**:
<img width="253" height="547" alt="Screenshot 2025-07-20 at 4 03 35 PM" src="https://github.com/user-attachments/assets/6b3b7e05-171a-49c5-a28d-fa1be04bd250" />

Food Search Setup screen. Pick the AI service you prefer, enter your own API keys:
<img width="251" height="548" alt="Screenshot 2025-07-20 at 4 03 46 PM" src="https://github.com/user-attachments/assets/2e864b44-8818-45c3-84e3-874d2fb0fd3d" />

taylorpatterson-T1D committed:
New Feature: Food Search 2.0 for Loop!

A new AI-enabled food search capability for Loop. Loop has helped me become a better diabetic and I would love to give back to the community by contributing this new feature for Loop, which I believe will benefit all Loop users.

PROBLEM WE'RE SOLVING: Diabetics often lack nutrition knowledge when estimating carb intake. And we need that knowledge when we’re making dosing decisions - this knowledge is the single most important tool we have to improve our Time in Range and A1C. Loop's Food Search feature uses barcodes and AI analysis to provide accurate nutrition information and advanced diabetes management recommendations.

NEW FEATURE'S IMPACT: The Food Search system is a comprehensive food analysis and nutrition tracking solution integrated into Loop for improved diabetes management. It provides multiple search options including text, barcode scanning, voice, and AI-powered image analysis that produces comprehensive nutrition data and dosing suggestions via diabetic notes. This let's us enter a meal bolus with MUCH greater accuracy to improve Time in Range and A1C. The AI-enabled analysis engine can even read menus (in multiple languages) and provide guidance for dosing.

Short video demo: https://youtu.be/L0LD8AxNX0Q

Review the docs in “/Loop/Documentation/FoodSearch 2.0 Docs/” for comprehensive documentation for Loop's Food Search functionality, including AI-powered nutrition analysis and advanced diabetes management recommendations. API setup guides are provided so the user can enter their personal AI API keys to enable AI food analysis.

This new PR (Food-Search 2.0) fixes compatibility bugs from the initial commit and adds Advanced AI Analysis including FPU analysis, and Dynamic Dosing recommendations.

Requesting a review from the maintainers? @ps2, @marionbarker, @loudnate

<I will post new screenshots here after submitting this PR>

taylorpatterson-T1D committed:
Delete Documentation/FoodSearch 2.0 Docs/Development 1.0 Docs directory

Deleting older docs

taylorpatterson-T1D committed:
Update project.pbxproj to ver 54

taylorpatterson-T1D committed:
Delete Loop/DefaultAssets.xcassets/AI-logo-master.png

not needed

taylorpatterson-T1D committed:
Added SupportsLiveActivities to Info.plist

To prevent breaking Live Activities customizations

taylorpatterson-T1D committed:
Merge branch 'dev' of https://github.com/taylorpatterson-T1D/Loop.Food-Search into dev

taylorpatterson-T1D committed:
Fix for build error on AbsorptionTimePickerRow and pbxproj file

Added xcodeproj/project.pbxproj to .gitignore
fixed build errors caused by AbsorptionTimePickerRow function

taylorpatterson-T1D committed:
Added delete food item Food Details list

- Red X delete buttons in Food Details so users can adjust what food items they plan to eat and not eat
- Automatic nutriments total & absorption time recalculations when items are deleted

taylorpatterson-T1D committed:
Delete Loop/Base.lproj/Main.storyboard

included accidentally so I'm removing it. should remain unchanged.

taylorpatterson-T1D committed:
Revert "Added SupportsLiveActivities to Info.plist"

This reverts commit 6f6569fe5ce26fe3a7abf93a284b30ea03721c8d.

taylorpatterson-T1D committed:
Moved isAbsorptionTimePickerRow change from /LoopKitUI back under /Loop

This change moves a change from /LoopKitUI senior directory down into the /LoopWorkSpace/Loop directory to keep everything under /Loop.

taylorpatterson-T1D committed:
BugFix for After AI scan, adjusting Servings does not update multi-circle nutriments.

taylorpatterson-T1D committed:
Adds a method to enable Food Search to Add Carb Entry View

  - Easy discovery: Users see the brain icon toggle when Food Search is disabled
  - Instant enablement: Toggle works immediately to show full Food Search UI
  - Consistent disabling: Settings changes (via gear icon) instantly update the UI
  - Responsive both ways: Works whether toggled from carb entry or settings

taylorpatterson-T1D committed:
AI Prompt performance improvement for Standard users

Split AI food analysis prompt into Standard and Advanced mode to increase performance when in Standard Mode.
  - Standard Mode: ~2,500 words (basic analysis only)
  - Advanced Mode: ~6,000+ words (includes FPU calculations, deeper food absorption details, exercise guidance, etc.)
  - Token Reduction: ~60% fewer tokens for Standard mode users

  What this means for users:
  - Standard mode users get faster AI analysis with streamlined prompts
  - Advanced mode users get comprehensive analysis when they need FPU calculations
  - The system automatically chooses the right prompt based on the Advanced Dosing setting

taylorpatterson-T1D committed:
Bugfix: Nutriments not accurately updated when deleting food items from AI analysis

Multi-Circle Nutriment Bug - Both food deletion and serving adjustments now work accurately and predictably.

taylorpatterson-T1D committed:
Refined AI prompts, added GPT 5 as option for AI.

1. Better Analysis Quality - GPT-5 provides improved accuracy for complex health related food analysis
2. Future-Proofed - Ready for GPT-5 when users have access
3. Cost-Effective Options - Users can choose between quality (GPT-5) and speed (GPT-4o-mini)

taylorpatterson-T1D committed:
Set search provider defaults to best ones at onboarding

Preferred food search providers for best performance are set to:
Text/Voice Search: USDA FoodData Central
Barcode Search: OpenFoodFacts
AI Image Analysis: OpenAI (ChatGPT API)

marionbarker commented:
## Code Review

I made some modifications to Taylor's code, starting at 477ca57, as documented in this branch:

See https://github.com/loopandlearn/Loop/commits/modify_for_pr2329
* 542a304e revert unintended changes to .gitignore file
* 2cce0efc restore xcworkspace and xcshareddata files, this change not related to this PR
* 68a738db Restore auto-signing capability, remove unneeded Libre references
* c5d4eaed Restore Main.storyboard from dev

Additional changes will be required, but the functionality can be tested.
* The additions (new files) use print statements instead of the log.default or log.debug
* I think the strings may require modifications before they can be localized, but I'll look into this.

More changes are found in a different branch that is focused on the minimum changes to support adding this as a customization: https://github.com/loopandlearn/Loop/commits/prepare_for_customization_pr2329
* Because automated tests for Loop are broken with the files added to LoopTests
   * I removed the LoopTests added in the PR (82dc481c)
   * I removed those file references in pbjproj file and commented out 2 test functions from `LoopTests/ViewModels/BolusEntryViewModelTests.swift` (ac3f87d6)
      *  testCarbEntryDateAndAbsorptionTimeString
      * testCarbEntryDateAndAbsorptionTimeString2
   * With those changes, Tests can now be run

## Test

Test with commit 542a304e.

This builds and appears to work. More testing will be needed.

marionbarker commented:
## Update with new commits

When Taylor updates his branch, then I need to also update my two branches:

In a local copy of Taylor's branch:
```
git pull
git switch modify_for_pr2329
git merge <local-name-for-taylor's-branch>
```

```
git switch prepare_for_customization_pr2329
git merge modify_for_pr2329
```

This keeps the 3 version in alignment.

taylorpatterson-T1D committed:
Fix width issue in the Add Carb Entry screen

The fix maintains the existing StackNavigationViewStyle() while adding proper width constraints. This should resolve the issue where the "Add Carb Entry" screen appears too wide and cuts off content after using AI analysis.

taylorpatterson-T1D committed:
Revert "Fix width issue in the Add Carb Entry screen"

This reverts commit a83d46d416ecd988b9886e8bd3f5c45368bc0b88.

taylorpatterson-T1D committed:
Fix width bug and rearrange Add Carb Entry view

The fix maintains the existing StackNavigationViewStyle() while adding proper width constraints. This should resolve the issue where the "Add Carb Entry" screen appears too wide and cuts off content after using AI analysis.

We also moved the Food Search enablement toggle and Food Search's section below Absorption Time row for a more logical flow.

Confirmed that the FPU formula implementation is CORRECT and adheres to the standard. No changes are needed to the calculation formula.

Minor tweak to AI prompt.

taylorpatterson-T1D committed:
Fixed camera capture misalignments

1. Perfected the Barcode window preview (white frame now bounds the scan window)
2. Perfected the cropping in AI picture capture (mis-cropped pics)

marionbarker commented:
On 2025-08-27, push changes to update the ancillary branches in loopandlearn that support the ability to add this feature as a customization to Loop. These updates bring the feature up to the level of `2c6a78e2` in this PR.

* this branch of Loop in loopandlearn GitHub username can be built with LoopWorkspace
   * `modify_for_pr2329` updated to `579d6cad`
* this branch of Loop in loopandlearn GitHub username is limited to only changes needed to support this feature
   * `prepare_for_customization_pr2329` updated to `d79db88e`

taylorpatterson-T1D committed:
Bugfixes - Favorite Foods and math fix for Carb Entry

1. bug-saving as favorite food does not actually save as favorite - Fixed
2. bug-math problem when calculating carbs for a portion using AI - Fixed

taylorpatterson-T1D committed:
Multiple feature UI improvements

-Add thumbnail to favorite foods lists
-Add Heart icon to action saving as Favorite Food
-Add Estimated Confidence percent when using AI
-When deleting food item show strikethrough font, add green plus, allow user to add itemback in

marionbarker commented:
I will be taking the last 2 commits one at a time.

As started earlier, I keep a parallel branch to handle building a customization from this PR.
* parallel branch is: https://github.com/loopandlearn/Loop/tree/modify_for_pr2329
* my local clone name for the branch in this PR is `taylor-try2-dev`

First: it looks like your commit 5aeda3b, Bugfixes - Favorite Foods and math fix for Carb Entry, inadvertently removed updates from two prior commits.

Please look at these commits:
* `bfd1c9d8` Fix width bug and rearrange Add Carb Entry view
   * https://github.com/taylorpatterson-T1D/Loop.Food-Search/commit/bfd1c9d8
* `97bb8d07` Set search provider defaults to best ones at onboarding
   * https://github.com/taylorpatterson-T1D/Loop.Food-Search/commit/97bb8d07

I believe both of those changes were desired and they are included in the Customization patch at the current time.

I was able to restore them using cherry-pick in my parallel branch, modify_for_pr2329.

```
git cherry-pick bfd1c9d8
git cherry-pick 97bb8d07
```
The code built at this point (with just the first commit from your branch merged in).

But when I pulled in the next commit and merged it then the code no longer built.

The next commit, `5cbb5d3c`, Multiple feature UI improvements, is not buildable when merged into my parallel branch, until I make one modification.

If I remove the thumbnail in this line: https://github.com/taylorpatterson-T1D/Loop.Food-Search/blob/5cbb5d3c/Loop/Views/FavoriteFoodsView.swift#L41, I am then able to build the code again.

The error is: 

> /Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/Loop/Loop/Views/FavoriteFoodsView.swift:21:25 The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions


This is with Xcode 16.4 building on macOS 15.6.1 building onto an iOS 18.6.2 phone.

Before I started today, my parallel branch was at 
* commit `579d6cad`, Merge branch 'taylor-try2-dev' into modify_for_pr2329, [link](https://github.com/loopandlearn/Loop/tree/579d6cad)

After doing the merges from the PR branch, the cherry-pick to restore desired changes and removing the one parameter that prevented me from building, the branch is at:
* commit `f7ec949f`, modify to enable building, [link](https://github.com/loopandlearn/Loop/tree/f7ec949f)

This is buildable by people who use automatic signing (and do not have Taylor's apple developer ID).

holger2de commented:
I tried to build it. I got the error message: No such module loopkit


Gesendet: Freitag, 29. August 2025 um 23:24
Von: "Marion Barker" ***@***.***>
An: LoopKit/Loop ***@***.***>
CC: Subscribed ***@***.***>
Betreff: Re: [LoopKit/Loop] New Feature: Food Search 2.0 for Loop! (PR #2329)

marionbarker left a comment (LoopKit/Loop#2329)
I will be taking the last 2 commits one at a time.
As started earlier, I keep a parallel branch to handle building a customization from this PR.

parallel branch is: https://github.com/loopandlearn/Loop/tree/modify_for_pr2329
my local clone name for the branch in this PR is taylor-try2-dev

First: it looks like your commit 5aeda3b, Bugfixes - Favorite Foods and math fix for Carb Entry, inadvertently removed updates from two prior commits.
Please look at these commits:

bfd1c9d8 Fix width bug and rearrange Add Carb Entry view

***@***.***


97bb8d07 Set search provider defaults to best ones at onboarding

***@***.***



I believe both of those changes were desired and they are included in the Customization patch at the current time.
I was able to restore them using cherry-pick in my parallel branch, modify_for_pr2329.
git cherry-pick bfd1c9d8
git cherry-pick 97bb8d07

The code built at this point (with just the first commit from your branch merged in).
But when I pulled in the next commit and merged it then the code no longer built.
The next commit, 5cbb5d3c, Multiple feature UI improvements, is not buildable when merged into my parallel branch, until I make one modification.
If I remove the thumbnail in this line: https://github.com/taylorpatterson-T1D/Loop.Food-Search/blob/5cbb5d3c/Loop/Views/FavoriteFoodsView.swift#L41, I am then able to build the code again.
The error is:

/Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/Loop/Loop/Views/FavoriteFoodsView.swift:21:25 The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions

This is with Xcode 16.4 building on macOS 15.6.1 building onto an iOS 18.6.2 phone.
Before I started today, this branch was at

commit 579d6cad, Merge branch 'taylor-try2-dev' into modify_for_pr2329, link

After doing the merges from the PR branch, the cherry-pick to restore desired changes and removing the one parameter that prevented me from building, the branch is at:

commit f7ec949f, modify to enable building, link

This is buildable by people who do not have Taylor's apple developer ID.
—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

marionbarker commented:
@holger2de 

please use these instructions to add this as a customization to Loop. This PR is not suitable to be added to Loop as is. 

https://github.com/loopandlearn/lnl-scripts/pull/82#issuecomment-3070502141

taylorpatterson-T1D committed:
FoodFinder Final Commit

1. Rebranded Food Search as "FoodFinder"
2. If using text search, once a simple food item is selected, feature also saves the mapped Emoji icon (if applicable) related to the food in the Favorite Foods (FoodType) record. AI Analyzed foods with thumbnails saves the thumbnail instead.
3. In FoodFinder Settings - reorganized settings elements for simplicity and ease of use including a multi-tab for AI configurator.
4. Add ability to "Bring your own" AI provider - your mileage may vary.
5. Optimized Standard prompt performance to make it faster, added switching logic to preserve mandatory prompt settings, speed up the Standard prompt, and refine the results from the Advanced Prompt.
6. Improved Portions & Servings - Better describes how it calculated the amounts of food that it sees and how those compare to USDA standard portions.
7. Many I/O performance and prompt accuracy improvements.
8. Updated documentation under /Loop/FoodFinder Docs

taylorpatterson-T1D committed:
delete file with test keys

taylorpatterson-T1D committed:
Revert "delete file with test keys"

This reverts commit f97bd48ab42f0eb15483db6cca5b88cb64eca4b7.

taylorpatterson-T1D committed:
Update BYOTestConfig.swift

removed key data in file. File is only used for internal testing. Can be deleted.

taylorpatterson-T1D committed:
Delete LoopTests/FoodSearchIntegrationTests.swift

Deleting unnecessary test file that breaks other tests.

taylorpatterson-T1D committed:
Update AISettingsView.swift

renamed settings view in FoodFinder

taylorpatterson-T1D committed:
UI/UX and prompt Improvements

-Surface the AI-generated absorption reasoning in a collapsible view so user
  understands why absorption time changed.
- Flag low-confidence or assumption-heavy results more prominently (maybe yellow badges).
- Provide a quick tip overlay in the AI camera flow explaining how to frame photos  (lighting, reference objects, etc.) to improve accuracy.
- Cleaned up settings view iconography.
- Improved prompt to inspect and report foreign language menu items translated into users device language settings
- Improved reasoning and explanations for absorption time adjustments.

taylorpatterson-T1D committed:
UI/UX and prompt Improvements

-Surface the AI-generated absorption reasoning in a collapsible view so user
  understands why absorption time changed.
- Flag low-confidence or assumption-heavy results more prominently (maybe yellow badges).
- Provide a quick tip overlay in the AI camera flow explaining how to frame photos  (lighting, reference objects, etc.) to improve accuracy.
- Cleaned up settings view iconography.
- Improved prompt to inspect and report foreign language menu items translated into users device language settings
- Improved reasoning and explanations for absorption time adjustments.

taylorpatterson-T1D committed:
WE ARE DONE! (FINAL FoodFinder COMMIT?)

Refactored for GPT-5 performance improvements:
-Image preprocessing.
-Prompt management.
-Transport pipeline.
-Caching/telemetry polish.

taylorpatterson-T1D committed:
Merge branch 'dev' into feature/AutoPresets

taylorpatterson-T1D committed:
Merge pull request #1 from taylorpatterson-T1D/feature/AutoPresets

Apply recent localization updates

SherlockInSpace commented:
Hi everyone, I'm trying to get this feature up and running so I can help out as an eager person with T1D ! 😄 
I am currently building with the Browser instructions, which has been working, but using this patch or even the branch introduces many compilation errors for me. They range from variables being used that are not defined anywhere, to Swift 6 compiler issues.
Has anyone else experience these issues using the Browser instructions ?

marionbarker commented:
This PR will not build. There are modifications required to this and to LoopKit. There is a preliminary customization (for now).

https://www.loopandlearn.org/loop-features-in-development/#food-search

tanabay2705 commented:
Hello, for about 3 weeks it has not worked with Gemini. I can only get it to work with Chatgpt. The error message is "Connection Interrupteur" Analyse failed
<img width="590" height="1278" alt="Capture d’écran   2025-10-17 à 17 01 53" src="https://github.com/user-attachments/assets/24d1d0d9-5d26-4668-8be5-58cc9185bccb" />
. Can you help me? Thank you

rolandet commented:
Google deprecated the Gemini 1.5 LLM and this customization feature was hard coded to use that version. So, until @taylorpatterson-T1D can refactor his code to make it more customizable on which LLM is used for each LLM provider this will remain broken.

mangle618 commented:
@tanabay2705 
I built using this customization last week and you can use the byo configuration for Gemini free. Here's the url: 
https://generativelanguage.googleapis.com/v1beta/openai/
Then put in your api key and a current version of the api

gemini-2.5-flash or gemini-2.5-pro

I hope this helps. I had the same issue.

rolandet commented:
EDIT: And, I just confirmed that this does indeed work,

EDIT: Okay, I see now, Gemini has a openai API compatible interface. 

@tanabay2705 What your stating makes no sense. Gemini and OpenAI are two entirely different LLMs made by different companies. Why would you place openai in a google Gemini URL?

marionbarker committed:
Live-Activity: short labels for display, long labels for description

marionbarker committed:
Enable autoscaling in Live Activity widget to prevent truncation

marionbarker committed:
update string catalog for Live Activity modifications

marionbarker committed:
improve large font display for Live Activity;
* add section headings;
* modify strings

taylorpatterson-T1D committed:
Merge remote-tracking branch 'upstream/dev' into dev

taylorpatterson-T1D committed:
Merge branch 'modify-live-activity' into dev

taylorpatterson-T1D committed:
Revert "Merge branch 'modify-live-activity' into dev"

This reverts commit b66dee38f68d8a788b326e7a493f86abc62521b7, reversing
changes made to e3cc01c50c3a8b432fe7f499033c097614ab074b.

taylorpatterson-T1D committed:
Merge branch 'LoopKit:dev' into dev

taylorpatterson-T1D commented:
This PR is closed - no further development will happen on this PR. 
Its replacement is [PR #2404](https://github.com/LoopKit/Loop/pull/2404)
